File: scala/leet_code/1544.scala
/** This module provides a solution to the problem of making a string "good" by removing all instances of bad characters.
  * A string is "good" if it doesn't have two adjacent characters that are the same, but one is uppercase and the other is lowercase.
  * The function `makeGood` is the main function in this module.
  */
object Solution {

  /** Function to make a string "good" by removing all instances of bad characters.
    * A string is "good" if it doesn't have two adjacent characters that are the same, but one is uppercase and the other is lowercase.
    *
    * @param s a String representing the input string to be processed.
    * @return a String representing the processed string that is "good".
    */
  def makeGood(s: String): String = {
    // Initialize a mutable stack to keep track of the characters in the string
    val stack = new scala.collection.mutable.Stack[Char]

    // Iterate over the characters in the string
    for (c <- s) {
      // If the stack is empty, push the current character onto the stack
      if (stack.isEmpty) {
        stack.push(c)
      } else {
        // If the top character on the stack is the same as the current character but has a different case
        if (stack.top.toLower == c.toLower && stack.top != c) {
          // Pop the top character from the stack
          stack.pop
        } else {
          // Push the current character onto the stack
          stack.push(c)
        }
      }
    }

    // Reverse the stack and convert it to a string
    stack.reverse.mkString
  }
}

File: java/leet_code/1544.java
import java.util.Stack;

/**
 * This class provides a solution for the problem of making a string good.
 * A string is good if it doesn't have two adjacent characters that are the same, but one is
 * uppercase and the other is lowercase.
 * The problem is solved by using a stack to keep track of the characters in the string.
 * If the current character and the top character of the stack are the same but have different
 * cases, the top character of the stack is removed.
 * Otherwise, the current character is pushed onto the stack.
 * Finally, the characters in the stack are popped and appended to a string builder, which is then
 * reversed and returned as a string.
 *
 * @author jerodg
 */
class Solution {

    /**
     * Makes a string good by removing adjacent characters that are the same but have different cases.
     *
     * @param s The input string. It is a string of characters, which can include both uppercase
     *          and lowercase letters.
     *
     * @return The good string. It is a string of characters, which can include both uppercase
     *         and lowercase letters.
     */
    public String makeGood(final String s) {
        // Initialize a stack to keep track of the characters in the string
        final Stack<Character> stack = new Stack<>();

        // Iterate over the characters in the string
        for (final char c : s.toCharArray()) {
            // If the stack is not empty and the current character and the top character of the stack are the same but have different cases,
            // remove the top character of the stack
            if (!stack.isEmpty() && 32 == Math.abs(stack.peek() - (int) c)) {
                stack.pop();
            } else {
                // Otherwise, push the current character onto the stack
                stack.push(c);
            }
        }

        // Initialize a string builder
        final StringBuilder sb = new StringBuilder();

        // Pop the characters from the stack and append them to the string builder
        while (!stack.isEmpty()) {
            sb.append(stack.pop());
        }

        // Reverse the string builder and return it as a string
        return sb.reverse().toString();
    }
}
